File: 60_Permutation_Sequence/getPermutation_v2.py
import time
class Solution(object):
    def getPermutation(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: str
        """
        nums = [str(i) for i in xrange(1, n + 1)]
        for i in xrange(k):
            self.nextPermutation(nums)
        return ''.join(nums)

    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        length = len(nums)
        index = length - 1
        while nums[index - 1] >= nums[index] and index:
            index -= 1

        if not index:
            nums.sort()
            return

        j = length - 1
        while j >= index - 1:
            if nums[j] > nums[index - 1]:
                nums[j], nums[index - 1] = nums[index - 1], nums[j]
                nums[index:] = sorted(nums[index:])
                break
            else:
                j -= 1

sol = Solution()
n = 9
k = 171669
start = time.time()
print sol.getPermutation(n, k)
print time.time() - start

File: 60_Permutation_Sequence/getPermutation_v2.cpp
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    string getPermutation(int n, int k) {
        string s = "";
        string res = "";
        for(int i = 1; i <= n; i++) {
            s.append(to_string(i));
        }
        dfs(res, s, k, 0);
        return res;
    }
    void dfs(string& res, string& s, int k, int startIndex) {
        if(startIndex >= s.size()) {
            k--;
        }
        
        if(k == 0) {
            res = s;
        }

        if(k < 0) {
            return;
        }

        for(int i = startIndex; i < s.size(); i++) {
            swap(s[i], s[startIndex]);
            dfs(res, s, k, startIndex + 1);
            swap(s[i], s[startIndex]);
        }
    }
};

int main() {
    Solution sol;
    int n = 3, k = 3;
    cout << sol.getPermutation(n, k) << endl;
}
